Fixed required attribute not being correctly recognized by form dialog

engineering-this opened the pull request:
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

I've changed form dialog to set its property `required = true` whenever matching attribute exists on form element.

Closes #586

Comandeer commented:
Please also rebase onto latest `major`.

Comandeer commented:
Changed base to `master` and rebased.

Comandeer commented on plugins/forms/plugin.js:
```diff
@@ -277,3 +277,16 @@ CKEDITOR.plugins.add( 'forms', {
+	* @private
+	* @param {CKEDITOR.dom.element} element Element which required attribute is checked.
+	* */
+	_setupFormDialogRequired: function( element ) {
```
This is incorrect implementation. According to the [HTML LS standard](https://html.spec.whatwg.org/multipage/input.html#the-required-attribute):

> The required attribute is a boolean attribute. When specified, the element is required.

[And further](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute):

> A number of attributes are boolean attributes. The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value.

Therefore the correct test is to check if the element has appropriate attribute (`element.hasAttribute( 'required' )`), as even incorrect values are treated as true.

Comandeer commented on plugins/forms/plugin.js:
```diff
@@ -277,3 +277,16 @@ CKEDITOR.plugins.add( 'forms', {
 	}
 } );
+
+CKEDITOR.plugins.forms = {
```
If we add new namespace, we should also add API docs for it.

Comandeer commented on tests/plugins/forms/manual/required.md:
```diff
@@ -0,0 +1,12 @@
+@bender-tags: editor, bug, 4.10.0, forms, 586
```
Please update version tag and remove redundant `forms` one.

engineering-this commented:
- Changed base to next.
- Changed check to `hasAttribute`, so either pressence or any value of 'required' attribute on form element will be reflected in dialog.
- Corrected docs entry.
- Refactored tests by extracting duplicate code to global helper.

Comandeer commented on plugins/forms/plugin.js:
```diff
@@ -277,3 +277,24 @@ CKEDITOR.plugins.add( 'forms', {
+	* @private
+	* @param {CKEDITOR.dom.element} element An element which required attribute is checked.
+	*/
+	_setupFormDialogRequired: function( element ) {
```
That name rubs me up the wrong way… Maybe something like `setupRequiredAttribute`?

Comandeer commented on plugins/forms/plugin.js:
```diff
@@ -277,3 +277,24 @@ CKEDITOR.plugins.add( 'forms', {
+ */
+CKEDITOR.plugins.forms = {
+	/**
+	* Sets dialog required value.
```
Maybe add some more info, e.g. it's based on algorithm described in specification.

Comandeer commented on tests/plugins/forms/manual/required.html:
```diff
@@ -0,0 +1,15 @@
+<body>
+<div id="editor">
+	<p>
+		<input required type="checkbox">
```
I'd be nice to have more than one `[required]` form, e.g. `[required=whatever]`, `[required=false]` etc.

Comandeer commented on CHANGES.md:
```diff
@@ -6,6 +6,7 @@
 Fixed Issues:
 
 * [#1181](https://github.com/ckeditor/ckeditor-dev/issues/1181): [Chrome] Fixed: Opening context menu in readonly editor results in error.
+* [#586](https://github.com/ckeditor/ckeditor-dev/issues/586) Fixed: required attribute not being correctly recognized by form dialog.
```
Please add reference to the original contributor.

Comandeer commented on tests/plugins/forms/_helpers/tools.js:
```diff
@@ -0,0 +1,25 @@
+		/**
+		 * Asserts 'required' attribute of a form element.
+		 *
+		 * @param {Object} setup Test setup object.
```
Why not simply `options`? We use this word everywhere else.

engineering-this commented:
Updated accordingly to requested changes, no changes to way anything works.
